doc/articles/Asynchronous Operations.md

```diff
@@ -0,0 +1,68 @@
+# Asynchronous Operations {#Asynchronous-Operations}
+
+Asynchronous operations in webgpu.h return a @ref WGPUFuture. This is an opaque handle which applications may use to poll or wait for completion of the asynchronous operation.
+
+On all implementations of webgpu.h, it must be possible to react to async operations without undue latency or spin-waiting.
+
+## Creation
+
+All asynchronous operations start when the application calls an asynchronous webgpu.h function. These functions return `WGPUFuture`, and take in their arguments a `CallbackInfo` struct like @ref WGPURequestDeviceCallbackInfo or @ref WGPUBufferMapCallbackInfo. `CallbackInfo` structs have the following members:
+ - a `WGPUChainedStruct const * nextInChain` pointer, for extensions.
+ - a @ref WGPUCallbackMode `mode` enum.
+ - a `callback` function pointer. For example:
+   ```typedef void (*WGPUBufferMapCallback)(
+        WGPUMapAsyncStatus status,
+        char const * message,
+        WGPU_NULLABLE void* userdata1,
+        WGPU_NULLABLE void* userdata2) WGPU_FUNCTION_ATTRIBUTE;```
+ - Two application-owned userdata members.\n
+   `void* userdata1`\n
+   `void* userdata2`
+
+The `callback` function pointer is called when the application _observes completion_ of the asynchronous operation. The `userdata1` and `userdata2` members are passed back to the application as the last two arguments in the callback function. Callbacks **may not** be called unless the application explicitly flushes them in order to _observe completion_. The point in time a callback is called depends on the @ref WGPUCallbackMode of the operation. webgpu.h provides three callback modes: `::WGPUCallbackMode_WaitAnyOnly`, `::WGPUCallbackMode_AllowProcessEvents`, and `::WGPUCallbackMode_AllowSpontaneous`.
+
+> @copydoc ::WGPUCallbackMode_WaitAnyOnly
+> @copydoc ::WGPUCallbackMode_AllowProcessEvents
+> @copydoc ::WGPUCallbackMode_AllowSpontaneous
+
+## wgpuInstanceWaitAny {#Wait-Any}
+`WGPUWaitStatus wgpuInstanceWaitAny(WGPUInstance, size_t futureCount, WGPUFutureWaitInfo * futures, uint64_t timeoutNS)`
+
+Waits on any WGPUFuture in the list of `futures` to complete for `timeoutNS` nanoseconds. Returns when at least one `WGPUFuture` is completed or `timeoutNS` elapses, whichever is first. If `timeoutNS` is zero, all `futures` are polled once, without blocking.
+
+Returns `::WGPUWaitStatus_Success` if at least one `WGPUFuture` completes. WGPUFutureWaitInfo::completed is set to true for all completed futures. See @ref WGPUWaitStatus for other status codes.
+
+Within this call, for any `WGPUFuture`s that completed, their respective callbacks will fire.
+
+### Timed Wait {#Timed-Wait}
+
+Use of _timed waits_ (`timeoutNS > 0`), must be enabled on the WGPUInstance with WGPUInstanceFeatures::timedWaitAnyEnable, and the number of futures waited on must be less than or equal to WGPUInstanceFeatures::timedWaitAnyMaxCount. Supported instance features may be queried using `::wgpuGetInstanceFeatures`.
```

Comment:
IIRC we also said that instance features would have to be enabled in wgpuCreateInstance (can leave a TODO since we don't have this in CreateInstance yet)

Reply:
we do have it, clarified further to say you have to enable it in the WGPUInstanceFeatures passed to wgpuCreateInstance